File: satellite/pipeline/engines/data_quality.py
"""Coverage-capped heuristic. No diagnostic accuracy or calibration claim."""
import json
import math
from pathlib import Path

POLICY = json.loads((Path(__file__).resolve().parents[3] / 'app/src/data/qualityPolicy.json').read_text())
MIN_VALID_PIXELS = POLICY['min_valid_pixels']

def data_quality_confidence(cloud_cover: float | None, pixel_count: int | None) -> float:
    if isinstance(pixel_count, bool) or not isinstance(pixel_count, (int, float)) or not math.isfinite(pixel_count):
        return 0.0
    if pixel_count != int(pixel_count) or pixel_count < MIN_VALID_PIXELS:
        return 0.0
    if isinstance(cloud_cover, bool) or not isinstance(cloud_cover, (int, float)) or not math.isfinite(cloud_cover) or not 0 <= cloud_cover <= 100:
        return 0.0
    fraction = min(1.0, pixel_count / POLICY['expected_pixels'])
    if fraction < POLICY['min_valid_fraction']:
        return 0.0
    weighted = POLICY['scene_weight'] * max(0.0, 100 - cloud_cover * POLICY['cloud_penalty']) + POLICY['pixel_weight'] * fraction * 100
    # Same positive-number rounding as the frontend; cap cannot exceed observed %.
    return math.floor(min(POLICY['max_score'], weighted, fraction * 100) * 10 + 0.5) / 10

File: app/src/lib/dataQuality.ts
import policy from '../data/qualityPolicy.json' with { type: 'json' };
export const MIN_VALID_PIXELS = policy.min_valid_pixels;
export type QualityInput = { cloud_cover?: number | null; pixel_count?: number | null; assessability?: string; observation_role?: string };
/** Heuristic image completeness, never probability of a correct diagnosis.
 * Zero means insufficient/unknown evidence. Scene clarity cannot replace local pixels.
 * Keep in sync with engines/data_quality.py through shared policy and parity tests.
 */
export function dataQualityConfidence(props: QualityInput): number {
  const cloud = props.cloud_cover, pixels = props.pixel_count;
  if (props.assessability === 'unassessable' || props.observation_role === 'retained_last_good') return 0;
  if (typeof pixels !== 'number' || !Number.isFinite(pixels) || !Number.isInteger(pixels) || pixels < MIN_VALID_PIXELS) return 0;
  if (typeof cloud !== 'number' || !Number.isFinite(cloud) || cloud < 0 || cloud > 100) return 0;
  const fraction = Math.min(1, pixels / policy.expected_pixels);
  if (fraction < policy.min_valid_fraction) return 0;
  const weighted = policy.scene_weight * Math.max(0, 100 - cloud * policy.cloud_penalty) + policy.pixel_weight * fraction * 100;
  return Math.round(Math.min(policy.max_score, weighted, fraction * 100) * 10) / 10;
}
